### PR TITLE
feat: get member clusters filtered by conditions

### PR DESCRIPTION
**Pull request by xcoulon**
filter by readiness and available capacity.
The readiness is found in the ClusterState, whereas
the capacity is found in the labels: a label with
key "toolchain.dev.openshift.com/capacity-exhausted"
is expected to be set to "true" when the member
cluster capacity has exhausted (setting this label
is out of the scope of this PR)

Updates #CRT-406

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (MatousJobanek) [inline .vscode/launch.json]: I guess that we don't need to add this file

File: pkg/cluster/cache.go [hunk 2]
```diff
@@ -62,12 +67,36 @@ func (c *kubeFedClusterClients) getFedCluster(name string) (*FedCluster, bool) {
...
+type Condition func(cluster *FedCluster) bool
+
+// Ready checks that the cluster is in a 'Ready' status condition
+var Ready Condition = func(cluster *FedCluster) bool {
+	for _, condition := range cluster.ClusterStatus.Conditions {
+		if condition.Type == common.ClusterReady && condition.Status == apiv1.ConditionTrue {
+			return true
+		}
+	}
+	return false
```
Comment (MatousJobanek) [inline pkg/cluster/cache.go]: you can replace the whole content of the function by:
```
return util.IsClusterReady(cluster.ClusterStatus)
```

File: pkg/cluster/cache_whitebox_test.go [hunk 3]
```diff
@@ -159,16 +159,65 @@ func TestGetClustersByType(t *testing.T) {
...
+			clusters := GetMemberClusters(Ready)
+
+			//then
+			assert.Len(t, clusters, 3)
+			assert.Contains(t, clusters, member1)
+			assert.Contains(t, clusters, member2)
+			assert.Contains(t, clusters, member4)
+		})
+
+		t.Run("get only available member clusters", func(t *testing.T) {
```
Comment (MatousJobanek) [inline pkg/cluster/cache_whitebox_test.go]:
```suggestion
		t.Run("get only member clusters with free capacity", func(t *testing.T) {
```

File: pkg/cluster/cache_whitebox_test.go [hunk 4]
```diff
@@ -159,16 +159,65 @@ func TestGetClustersByType(t *testing.T) {
...
+			clusters := GetMemberClusters(CapacityNotExhausted)
+
+			//then
+			assert.Len(t, clusters, 3)
+			assert.Contains(t, clusters, member1)
+			assert.Contains(t, clusters, member2)
+			assert.Contains(t, clusters, member3)
+		})
+
+		t.Run("get only ready and available member clusters", func(t *testing.T) {
```
Comment (MatousJobanek) [inline pkg/cluster/cache_whitebox_test.go]:
```suggestion
		t.Run("get only ready member clusters that have free capacity", func(t *testing.T) {
```

File: .vscode/launch.json (hunk 1 above)
Comment (xcoulon) [reply .vscode/launch.json]: ah right, it was not meant to be committed 🤦‍♂

File: pkg/cluster/cache.go (hunk 2 above)
Comment (xcoulon) [reply pkg/cluster/cache.go]: ok, will do

File: .vscode/launch.json (hunk 1 above)
Comment (xcoulon) [reply .vscode/launch.json]: remove in 41d5fe2

**Comment by xcoulon** [comment]
> I like the way of filtering the clusters 👍

thanks!

File: pkg/cluster/cache.go (hunk 2 above)
Comment (xcoulon) [reply pkg/cluster/cache.go]: done in 1222e2d

**Pull request merged**